x/poolmanager/client/query_proto_wrap.go

```diff
@@ -254,3 +256,162 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// If TwapPrice is specified we need to adjust the maxPriceImpact based on the deviation between spot and twap.
+	adjustedMaxPriceImpact := req.MaxPriceImpact
+	if !req.TwapPrice.IsZero() {
+		priceDeviation := spotPrice.Sub(req.TwapPrice).Quo(req.TwapPrice)
+		adjustedMaxPriceImpact = adjustedMaxPriceImpact.Sub(priceDeviation)
+
+		// If the adjusted max price impact is negative or zero it means the difference between spot price and
+		// twap already exceeds the max price impact.
+		if adjustedMaxPriceImpact.IsZero() || adjustedMaxPriceImpact.IsNegative() {
+			return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+				InputCoin:  sdk.NewCoin(req.FromCoin.Denom, sdk.ZeroInt()),
+				OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+			}, nil
+		}
+	}
+
+	// First, try the full 'from coin' amount without a swap fee, if it errors it means there
+	// is very low liquidity for the trade we are attempting.
+	tokenOut, err := swapModule.CalcOutAmtGivenIn(ctx, poolI, req.FromCoin, req.ToCoinDenom, sdk.ZeroDec())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// If tokenOut is zero it means we are trying to trade dust.
+	if tokenOut.IsZero() {
+		return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+			InputCoin:  req.FromCoin,
+			OutputCoin: tokenOut,
+		}, nil
+	}
+
+	currTradePrice := sdk.NewDec(req.FromCoin.Amount.Int64()).QuoInt(tokenOut.Amount)
+	priceDeviation := currTradePrice.Sub(spotPrice).Quo(spotPrice).Abs()
+
+	if priceDeviation.LTE(adjustedMaxPriceImpact) {
+		// If the full 'from coin' amount results in a price deviation less than or equal to the adjusted max price
+		// impact, return it
+
+		// Re-calculate the tokenOut with a swap fee.
+		tokenOut, err = swapModule.CalcOutAmtGivenIn(ctx, poolI, req.FromCoin, req.ToCoinDenom, poolI.GetSpreadFactor(ctx))
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+			InputCoin:  req.FromCoin,
+			OutputCoin: tokenOut,
+		}, nil
+	}
+
+	// Define low and high amount to search between. Start from 1 and req.FromCoin.Amount as initial range.
+	lowAmount := sdk.OneInt()
+	highAmount := req.FromCoin.Amount
+	currFromCoin := req.FromCoin
+
+	for lowAmount.LTE(highAmount) {
+		// Calculate middle amount
+		midAmount := lowAmount.Add(highAmount).Quo(sdk.NewInt(2))
+
+		// Update currFromCoin
+		currFromCoin = sdk.NewCoin(req.FromCoin.Denom, midAmount)
+
+		// If we error in the input it means we've gotten too low and are trading dust.
+		tokenOut, err := swapModule.CalcOutAmtGivenIn(
+			ctx, poolI, currFromCoin, req.ToCoinDenom, sdk.ZeroDec(),
+		)
+		if err != nil || tokenOut.IsZero() {
+			return &queryproto.EstimateTradeBasedOnPriceImpactResponse{
+				InputCoin:  req.FromCoin,
+				OutputCoin: sdk.NewCoin(req.ToCoinDenom, sdk.ZeroInt()),
+			}, nil
+		}
```

Comment:
If we've gotten too low, shouldn't we set this value as the new low?

In other words... if 7 is the correct answer and we're searching 1-10 (mid=5), where 5 is too low, shouldn't we still try 5-10?

Reply:
I don't think so, we've tried multiple high amounts to the point of trying dust even if we increase it a bit it's still dust. 5-10 would be dust so there is practically no tradeable amount or even worth the gas fee to do so.

Reply:
I think it's also overcomplicating it, I feel like having it error here the algo tries "the best it can" to find a tradeable amount, we don't need a perfect value.

Reply:
But we haven't checked 5-10 at this stage no? How would you know they are dust as well?

Reply:
if the first CalcOutAmtGivenIn works (outside the binary search) but the CalcOutAmtGivenIn in the first iteration fails (half fromCoin), there is a value where the trade is valid